src/rules/array.jl

```diff
@@ -0,0 +1,56 @@
+#####
+##### `reshape`
+#####
+
+function rrule(::typeof(reshape), A::AbstractArray, dims::Tuple{Vararg{Int}})
+    return reshape(A, dims), (Rule(Ȳ->reshape(Ȳ, dims)), DNERule())
+end
+
+function rrule(::typeof(reshape), A::AbstractArray, dims::Int...)
+    Y, (rule, _) = rrule(reshape, A, dims)
+    return Y, (rule, fill(DNERule(), length(dims))...)
+end
+
+#####
+##### `hcat` (🐈)
+#####
+
+function rrule(::typeof(hcat), A::AbstractArray, Bs::AbstractArray...)
+    Y = hcat(A, Bs...)
+    Xs = (A, Bs...)
+    rules = ntuple(length(Bs) + 1) do i
+        l = mapreduce(j->size(Xs[j], 2), Base.add_sum, 1:i-1; init=0)
+        u = l + size(Xs[i], 2)
+        dim = u > l + 1 ? (l+1:u) : u
+        Rule(Ȳ->copy(selectdim(Ȳ, 2, dim)))
```

Comment:
why `copy`?
For my education on how ChainRules thinks of defensive copies

Reply:
Definitely a defensive copy in this case; `selectdim` returns a view, which `copy` materializes. This is the same thing that Nabla does.

Reply:
Right, I see why that is needed.

Thanks

Reply:
I should add a comment to that effect, so thanks for asking. :slightly_smiling_face: